Prose/ProseLib/Sources/SharedModels/CredentialsClient.swift

```diff
@@ -0,0 +1,91 @@
+//
+//  CredentialsClient.swift
+//  Prose
+//
+//  Created by Rémi Bardon on 15/06/2022.
+//
+
+import Foundation
+
+/// - Note: See <https://developer.apple.com/documentation/security/keychain_services/keychain_items/using_the_keychain_to_manage_user_secrets>
+///         for documentation about the Keychain.
+/// - Copyright: Inspired by <https://gist.github.com/nesium/4a3da805d6000b76350cbe9c8aa35cf2>.
+public struct CredentialsClient {
+    public var loadCredentials: (_ jid: JID) throws -> String?
+    public var save: (_ jid: JID, _ password: String) throws -> Void
+    public var deleteCredentials: (_ jid: JID) throws -> Void
+}
+
+private enum KeychainError: Error {
+    case unexpectedPasswordData
+    case unhandledError(status: OSStatus)
+}
+
+public extension CredentialsClient {
```

Comment:
When I added the `SharedModels` package my intention was to bundle [domain models](https://martinfowler.com/eaaCatalog/domainModel.html). The `CredentialsClient` and `UserDefaultsClient` are dependencies however.

In another project I've added a package `Toolbox` put almost all dependencies in there and called it a day. The pointfree people add a package for each dependency + dependency "live". That seemed a bit much to me. Yesterday however I saw a great talk that also proposed having "interface packages" (just the dependency but not the "live" implementation) since it can help with the parallelization of the build. Multiple packages would depend on the interface only which is quick to build and are not blocked by the longer compilation time of the implementation.

Thus I'd say let's bite the bullet and add a package for each dependency. But also put the live implementation in there for now. We can later pull out the implementation if we need to and would only have to change the one site which builds up the `AppEnvironment`.

Reply:
Yeah, I dind't like putting it in `SharedModels` either, so I'll do this.

So we're talking about a `CredentialsClient` and a `UserDefaultsClient` **targets**? Or are you really talking about Swift **packages**?

Reply:
Sorry, **targets**. You're correct.